### DISCUSSION TITLE
Is it possible to list workflows of a repository using the GraphQL API?

### DISCUSSION THREAD

**Discussion by gaborcsardi** [community/community#56300] [Apps, API and Webhooks] [answered]
### Select Topic Area

Question

### Body

I am trying to find a way to list the workflows of a repository using the GraphQL API and/or to get a workflow of a repo using its file name. The REST API has an endpoints for this: https://docs.github.com/en/rest/actions/workflows?apiVersion=2022-11-28#list-repository-workflows
https://docs.github.com/en/rest/actions/workflows?apiVersion=2022-11-28#get-a-workflow

But seemingly the GraphQL API does not have a connection from `repository` to `workflows`. Am I missing something?

**Comment by LangLangBart** [comment, marked as answer]
**NOTE:** GitHub’s `REST API` and `GraphQL API` have no parity, they are a Venn Diagram, feature-wise.
- Request for a GraphQL Actions API is quite old already #24493

---

As a workaround, you could peek into the `.github/workflows/` folder and see its contents
- see the [git-scm.com/docs/git-rev-parse](https://git-scm.com/docs/git-rev-parse) for some valid `expressions` one can run on it, e.g. `"trunk@{3 years 20 weeks 3 days 1 hour 1 second ago}:.github/workflows/"`

```graphql
query ($owner: String = "cli", $name: String = "cli") {
  repository(owner: $owner, name: $name) {
    nameWithOwner
    object(expression: "trunk:.github/workflows/") {
      ... on Tree {
        entries {
          name
          object {
            ... on Blob {
              byteSize
              # text
            }
          }
        }
      }
    }
  }
}
```
```json
{
  "data": {
    "repository": {
      "nameWithOwner": "cli/cli",
      "object": {
        "entries": [
          {
            "name": "codeql.yml",
            "object": {
              "byteSize": 756
            }
          },
          {
            "name": "deployment.yml",
            "object": {
              "byteSize": 12548
            }
          },
          {
            "name": "go.yml",
            "object": {
              "byteSize": 817
            }
          },
          {
            "name": "issueauto.yml",
            "object": {
              "byteSize": 605
            }
          },
          {
            "name": "lint.yml",
            "object": {
              "byteSize": 1708
            }
          },
          {
            "name": "prauto.yml",
            "object": {
              "byteSize": 3366
            }
          }
        ]
      }
    }
  }
}
```

**Reply by gaborcsardi** [reply]
Thanks for your reply!

> Request for a GraphQL Actions API is quite old already https://github.com/orgs/community/discussions/24493

Yeah, there is some support for actions, actually, even for workflows. E.g. if I know the node id of a workflow, then I can query it. But there doesn't seem to be a link from the repository to the workflows, at least neither of us could find it.

> As a workaround, you could peek into the `.github/workflows/` folder and see its contents

Yeah, that might be sufficient in some cases, but a file can be there and actions might still be not running, e.g. for forked repos by default. The other issue is that I want to know if the workflow is enabled or disabled.

Thanks again for your reply, I'll take it as a reinforcement that currently there is no repo -> workflows link in GraphQL, and I'll need to use the REST API.

**Reply by LangLangBart** [reply]
> Yeah, there is some support for actions, actually, even for workflows. E.g. if I know the node id of a workflow, then I can query it. But there doesn't seem to be a link from the repository to the workflows, at least neither of us could find it.

You could go through the `commits` but that would only show you some workflows, you wouldn't be able to get a complete picture of all existing workflows and whether they are `active` or disabled.

```graphql
query ($owner: String = "cli", $name: String = "cli") {
  repository(owner: $owner, name: $name) {
    nameWithOwner
    defaultBranchRef {
      target {
        ... on Commit {
          checkSuites(last: 1) {
            nodes {
              conclusion
              workflowRun {
                workflow {
                  name
                  state
                }
              }
            }
          }
        }
      }
    }
  }
}
```

```json
{
  "data": {
    "repository": {
      "nameWithOwner": "cli/cli",
      "defaultBranchRef": {
        "target": {
          "checkSuites": {
            "nodes": [
              {
                "conclusion": "SUCCESS",
                "workflowRun": {
                  "workflow": {
                    "name": "Tests",
                    "state": "ACTIVE"
                  }
                }
              }
            ]
          }
        }
      }
    }
  }
}
```

**Reply by gaborcsardi** [reply]
Yeah, exactly. Every link seems to go through `checkSuites`, and then it will miss workflows that haven't run yet, or are disabled, etc.
